source/async_stack.cpp

```diff
@@ -145,6 +149,20 @@ static thread_local AsyncStackRootHolder currentThreadAsyncStackRoot;
 
 }  // namespace
 
+namespace utils {
+  uint64_t getOSThreadID() {
+   #if defined(__APPLE__)
+     uint64_t tid;
+     pthread_threadid_np(nullptr, &tid);
+     return tid;
+   #elif defined(_WIN32)
+     return uint64_t(GetCurrentThreadId());
+   #else
+    return uint64_t(syscall(UNIFEX_SYS_gettid));
```

Comment:
Can we avoid defining `UNIFEX_SYS_gettid` by just switching here on whether `SYS_gettid` is defined?

Also, I'm a bit anxious about the cost of a syscall in each `AsyncStackRoot` constructor since we construct so many of them. Is this a cheap syscall implemented in user space? If so then it doesn't matter, but if it's a typical syscall that transits into kernel space then we should probably revisit capturing this value in the `AsyncStackRootHolder` since we'll have one of those per thread and the per-thread value of this is constant.

Reply:
So I updated this to use the glibc wrapper of gettid because it implements caching ([D49174906](https://www.internalfb.com/diff/D49174906))

I also spoke to David and his response is that it's difficult to say because it depends on the implementation and we need to profile the code in order to know. This will be used a lot on android and his response is below:

> I see. The overhead of the syscall is mostly dependent on what speculative execution mitigations are turned on e.g. is the kernel hardened. I don't know how we configure our Android kernels.

The more mitigations the costlier the overhead of a user to kernel transition

The only real way to know is to profile the code. But as a hunch thread creation should be off the hot path so I don't imagine adding a gettid in there will be problematic

Reply:
Synced with @ispeters offline and decided to move threadid to async stack root holder for getting the threadid per thread only